src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs

```diff
@@ -404,24 +429,35 @@ private void CancelAfter(uint millisecondsDelay)
             // expired and Disposed itself).  But this would be considered bad behavior, as
             // Dispose() is not thread-safe and should not be called concurrently with CancelAfter().
 
-            TimerQueueTimer? timer = _timer;
+            ITimer? timer = _timer;
             if (timer == null)
             {
                 // Lazily initialize the timer in a thread-safe fashion.
                 // Initially set to "never go off" because we don't want to take a
                 // chance on a timer "losing" the initialization and then
                 // cancelling the token before it (the timer) can be disposed.
                 timer = new TimerQueueTimer(s_timerCallback, this, Timeout.UnsignedInfinite, Timeout.UnsignedInfinite, flowExecutionContext: false);
-                TimerQueueTimer? currentTimer = Interlocked.CompareExchange(ref _timer, timer, null);
+                ITimer? currentTimer = Interlocked.CompareExchange(ref _timer, timer, null);
                 if (currentTimer != null)
                 {
                     // We did not initialize the timer.  Dispose the new timer.
-                    timer.Close();
+                    timer.Dispose();
                     timer = currentTimer;
                 }
             }
 
-            timer.Change(millisecondsDelay, Timeout.UnsignedInfinite, throwIfDisposed: false);
+            if (timer is TimerQueueTimer tqt)
+            {
+                tqt.Change(millisecondsDelay, Timeout.UnsignedInfinite, throwIfDisposed: false);
+            }
+            else
+            {
+                try
+                {
+                    timer.Change(TimeSpan.FromMilliseconds(millisecondsDelay), Timeout.InfiniteTimeSpan);
+                }
+                catch (ObjectDisposedException) { }
```

Comment:
Three things occur to me here:
1) We're ignoring the return value of Change.  If Change can legitimately return false, and the timer wasn't impacted, then the CancelAfter call is possibly broken.  We should consider throwing an exception if Change returns false, e.g.
```C#
if (!timer.Change(...) && !IsCancellationRequested) throw ...
```
2) We could say that Change shouldn't throw exceptions, e.g. if the timer is already disposed, Change should just return false, in which case we could remove the try/catch.
3) We could consider changing the behavior of Timer.Change itself according to (2), such that rather than throwing an ODE it returns false.  That code path already exists (see the throwIfDisposed: false above), so we'd effectively just be making that the only behavior.  cc: @kouvel

Reply:
I am on the border line here. Not throwing is hiding some problem in the user code. Users sometime can handle the timer lifetime wrongly assuming the timer object is tracked while it is not and can be collected. In such cases I guess throwing will be more beneficiary.